perception/autoware_detected_object_validation/src/lanelet_filter/debug.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright 2022 TIER IV, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "lanelet_filter.hpp"
+
+#include <boost/geometry/algorithms/convex_hull.hpp>
+#include <boost/geometry/algorithms/disjoint.hpp>
+#include <boost/geometry/algorithms/intersects.hpp>
+#include <boost/geometry/index/rtree.hpp>
+
+#include <lanelet2_core/geometry/BoundingBox.h>
+#include <lanelet2_core/geometry/Lanelet.h>
+#include <lanelet2_core/geometry/LaneletMap.h>
+#include <lanelet2_core/geometry/Polygon.h>
+
+#include <optional>
+
+namespace autoware::detected_object_validation
+{
+namespace lanelet_filter
+{
+
+using visualization_msgs::msg::Marker;
+using visualization_msgs::msg::MarkerArray;
+
+template <typename T>
+std::optional<visualization_msgs::msg::Marker> createPolygonMarker(
+  T polygon, rclcpp::Time stamp, std::string_view ns, size_t marker_id,
+  std_msgs::msg::ColorRGBA color)
+{
+  if (polygon.empty()) {
+    return std::nullopt;
+  }
+  const constexpr std::string_view FRAME_ID = "map";
+
+  auto create_marker = [&](auto marker_type) {
+    Marker marker;
+    marker.id = marker_id;
+    marker.header.frame_id = FRAME_ID;
+    marker.header.stamp = stamp;
+    marker.ns = ns;
+    marker.type = marker_type;
+    marker.action = Marker::ADD;
+    marker.lifetime = rclcpp::Duration::from_seconds(0.1);
+    return marker;
+  };
+
+  auto marker = create_marker(Marker::LINE_LIST);
+  marker.pose.orientation.w = 1.0;
+  marker.scale.x = 0.1;
+  marker.color.a = color.a;
+  marker.color.r = color.r;
+  marker.color.g = color.g;
+  marker.color.b = color.b;
+
+  std::vector<geometry_msgs::msg::Point> points;
+  points.reserve(polygon.size());
+
+  for (size_t i = 0; i < polygon.size() - 1; ++i) {
+    geometry_msgs::msg::Point point;
+    point.x = polygon[i].x();
+    point.y = polygon[i].y();
+    points.push_back(point);
+
+    point.x = polygon[i + 1].x();
+    point.y = polygon[i + 1].y();
+    points.push_back(point);
+  }
```

Comment:
```suggestion
  std::vector<geometry_msgs::msg::Point> points;
  points.reserve(polygon.size() * 2);
  
  for (auto it = polygon.begin(); it != std::prev(polygon.end()); ++it) {
    geometry_msgs::msg::Point point;
    
    point.x = it->x();
    point.y = it->y();
    points.push_back(point);
    
    const auto next = std::next(it);
    point.x = next->x();
    point.y = next->y();
    points.push_back(point);
  }
```
@sebekx 
- reserved size looks like wrong.
- The iterator-based approach (it != std::prev(polygon.end())) provides better safety and container independence compared to index-based access.

Reply:
thanks, it's reasonable